aaelf64-morello/aaelf64-morello.rst

```diff
@@ -827,3 +844,91 @@ veneer is used. The BX changes the execution state from A64 to C64:
   adrp c16, sym
   add c16, c16, :lo12:sym
   br c16
+
+TLS for the pure capability ABI
+-------------------------------
+
+The design is based on TLSDESC, with the purpose of minimizing the performance
+differences between A64 and C64, while providing strict bounds when resolving
+TLS globals. Only the General Dynamic access model and the relaxation from
+General Dynamic to Local Exec are currently specified. However, this leaves the
+opportunity to add support for the other access models in the future.
+
+TLS static block
+^^^^^^^^^^^^^^^^
+
+The static block layout is the same used in AArch64 (Variant 1, see [TLS_]), with
+the only exception that TCB and the DTV pointer are capabilities.
+
+Thread pointer
+^^^^^^^^^^^^^^
+
+The thread pointer is a capability, held in ``CTPIDR_EL0``. The thread pointer
+needs to have the read, write, read capability and write capability permissions
+and bounds such that the entire TLS static block is accessible.
+
+Resolver functions
+^^^^^^^^^^^^^^^^^^
+
+A resolver function takes arguments in c0 (address of the TLS GOT slot), and c2
```

Comment:
This is an extremely weird calling convention. I can see _why_ you've done it (so that ctpidr_el0 can be read into c2 once and preserved), but I do not think that is a worthwhile optimisation given how little TLS is used when it makes such a confusing calling convention.

Reply:
The other reason is that if we would read CTPIDR in the resolver then we would also need to read it when doing the GD->LE relaxation (in the LE sequence), and that would add another NOP to the GD sequence.

Reply:
Now that x1 is gone as a return value you can instead pass CTPIDR in c1 at least and avoid the hole

Reply:
We still need a scratch register though. That could be c2, but then that leaves a gap in the preserved registers, which is less nice. I would rather leave this like it is (also avoids some large changes in both the spec and implementations).